File: Recomendacoes.js
class Recomendacoes {
    constructor() {
        this.categorias = {
            'NOTEBOOKS': p => p.nome.includes('Notebook'),
            'SMARTPHONES': p => p.nome.includes('iPhone') || p.nome.includes('Samsung') || p.nome.includes('Xiaomi') || p.nome.includes('Motorola'),
            'TABLETS': p => p.nome.includes('iPad') || p.nome.includes('Tab'),
            'SMART TVs': p => p.nome.includes('"'),
            'FONES DE OUVIDO': p => p.nome.includes('AirPods') || p.nome.includes('Sony') || p.nome.includes('Buds') || p.nome.includes('JBL'),
            'MONITORES': p => (p.nome.includes('Dell') || p.nome.includes('LG') || p.nome.includes('Samsung') || p.nome.includes('AOC')) && p.nome.includes('"'),
            'TECLADOS E MOUSES': p => p.nome.includes('Logitech') || p.nome.includes('Razer') || p.nome.includes('HyperX'),
            'IMPRESSORAS': p => p.nome.includes('HP') || p.nome.includes('Epson') || p.nome.includes('Canon') || p.nome.includes('Brother')
        };
    }

    encontrarCategoria(produto) {
        for (const [categoria, filtro] of Object.entries(this.categorias)) {
            if (filtro(produto)) {
                return categoria;
            }
        }
        return null;
    }

    encontrarCategoriasVisualizadas(historico) {
        const categoriasVisualizadas = new Set();
        historico.forEach(produto => {
            const categoria = this.encontrarCategoria(produto);
            if (categoria) {
                categoriasVisualizadas.add(categoria);
            }
        });
        return categoriasVisualizadas;
    }

    gerarRecomendacoes(produtosPrincipais, historico, carrinho) {
        const produtosVistos = new Set([...historico, ...carrinho].map(p => p.nome));
        const categoriasVisualizadas = this.encontrarCategoriasVisualizadas(historico);

        if (categoriasVisualizadas.size === 0) {
            return [];
        }

        let recomendacoes = [];
        for (const categoria of categoriasVisualizadas) {
            const produtosCategoria = produtosPrincipais
                .filter(produto => this.categorias[categoria](produto) && !produtosVistos.has(produto.nome))
                .map(produto => ({
                    ...produto,
                    categoria: categoria,
                    preco: produto.preco * (0.8 + Math.random() * 0.4)
                }));
            
            recomendacoes.push(...produtosCategoria);
        }

        return recomendacoes
            .sort(() => Math.random() - 0.5)
            .slice(0, 4);
    }

    agruparPorCategoria(recomendacoes) {
        const recomendacoesPorCategoria = {};
        recomendacoes.forEach(produto => {
            if (!recomendacoesPorCategoria[produto.categoria]) {
                recomendacoesPorCategoria[produto.categoria] = [];
            }
            recomendacoesPorCategoria[produto.categoria].push(produto);
        });
        return recomendacoesPorCategoria;
    }
}

// Exporta a classe para uso em outros arquivos
export default Recomendacoes; 

File: Recomendacoes.java
import java.util.*;
import java.util.function.Predicate;

public class Recomendacoes {
    private final List<Produto> produtosRecomendados;
    private final Map<String, Predicate<Produto>> categorias;

    public Recomendacoes() {
        this.produtosRecomendados = new ArrayList<>();
        this.categorias = new HashMap<>();
        inicializarCategorias();
    }

    private void inicializarCategorias() {
        categorias.put("NOTEBOOKS", p -> p.getNome().contains("Notebook"));
        categorias.put("SMARTPHONES", p -> p.getNome().contains("iPhone") || 
                                         p.getNome().contains("Samsung") || 
                                         p.getNome().contains("Xiaomi") || 
                                         p.getNome().contains("Motorola"));
        categorias.put("TABLETS", p -> p.getNome().contains("iPad") || 
                                     p.getNome().contains("Tab"));
        categorias.put("SMART TVs", p -> p.getNome().contains("\""));
        categorias.put("FONES DE OUVIDO", p -> p.getNome().contains("AirPods") || 
                                             p.getNome().contains("Sony") || 
                                             p.getNome().contains("Buds") || 
                                             p.getNome().contains("JBL"));
        categorias.put("MONITORES", p -> (p.getNome().contains("Dell") || 
                                        p.getNome().contains("LG") || 
                                        p.getNome().contains("Samsung") || 
                                        p.getNome().contains("AOC")) && 
                                        p.getNome().contains("\""));
        categorias.put("TECLADOS E MOUSES", p -> p.getNome().contains("Logitech") || 
                                                p.getNome().contains("Razer") || 
                                                p.getNome().contains("HyperX"));
        categorias.put("IMPRESSORAS", p -> p.getNome().contains("HP") || 
                                         p.getNome().contains("Epson") || 
                                         p.getNome().contains("Canon") || 
                                         p.getNome().contains("Brother"));
    }

    public void gerarRecomendacoes(List<Produto> historico, List<Produto> carrinho) {
        produtosRecomendados.clear();
        Set<String> produtosVistos = new HashSet<>();
        Set<String> categoriasVisualizadas = new HashSet<>();

        // Coleta produtos vistos e suas categorias
        for (Produto produto : historico) {
            produtosVistos.add(produto.getNome());
            String categoria = encontrarCategoria(produto);
            if (categoria != null) {
                categoriasVisualizadas.add(categoria);
            }
        }
        for (Produto produto : carrinho) {
            produtosVistos.add(produto.getNome());
        }

        // Gera recomendações por categoria
        if (!categoriasVisualizadas.isEmpty()) {
            for (String categoria : categoriasVisualizadas) {
                List<Produto> produtosCategoria = Produto.getProdutos().stream()
                    .filter(p -> !produtosVistos.contains(p.getNome()))
                    .filter(categorias.get(categoria))
                    .map(p -> criarProdutoRecomendado(p, categoria))
                    .toList();
                produtosRecomendados.addAll(produtosCategoria);
            }
        }

        // Adiciona produtos similares e complementares
        for (Produto produto : historico) {
            adicionarProdutosSimilares(produto);
        }
        for (Produto produto : carrinho) {
            adicionarProdutosComplementares(produto);
        }

        // Remove duplicatas e limita a quantidade
        produtosRecomendados.removeAll(carrinho);
        if (produtosRecomendados.size() > 4) {
            Collections.shuffle(produtosRecomendados);
            produtosRecomendados.subList(4, produtosRecomendados.size()).clear();
        }
    }

    private String encontrarCategoria(Produto produto) {
        for (Map.Entry<String, Predicate<Produto>> entry : categorias.entrySet()) {
            if (entry.getValue().test(produto)) {
                return entry.getKey();
            }
        }
        return null;
    }

    private Produto criarProdutoRecomendado(Produto original, String categoria) {
        double variacaoPreco = original.getPreco() * (0.8 + Math.random() * 0.4);
        return new Produto(
            original.getNome() + " (Recomendado)",
            "Produto recomendado da categoria " + categoria,
            variacaoPreco
        );
    }

    private void adicionarProdutosSimilares(Produto produto) {
        double precoVariacao = produto.getPreco() * 0.2;
        Produto similar = new Produto(
            produto.getNome() + " (Similar)",
            "Produto similar a " + produto.getNome(),
            produto.getPreco() + (Math.random() * precoVariacao * 2 - precoVariacao)
        );
        if (!produtosRecomendados.contains(similar)) {
            produtosRecomendados.add(similar);
        }
    }

    private void adicionarProdutosComplementares(Produto produto) {
        Produto complementar = new Produto(
            "Complemento para " + produto.getNome(),
            "Produto complementar a " + produto.getNome(),
            produto.getPreco() * 0.5
        );
        if (!produtosRecomendados.contains(complementar)) {
            produtosRecomendados.add(complementar);
        }
    }

    public Map<String, List<Produto>> agruparPorCategoria() {
        Map<String, List<Produto>> recomendacoesPorCategoria = new HashMap<>();
        for (Produto produto : produtosRecomendados) {
            String categoria = encontrarCategoria(produto);
            if (categoria != null) {
                recomendacoesPorCategoria.computeIfAbsent(categoria, k -> new ArrayList<>())
                    .add(produto);
            }
        }
        return recomendacoesPorCategoria;
    }

    public List<Produto> getProdutosRecomendados() {
        return new ArrayList<>(produtosRecomendados);
    }

    public void limparRecomendacoes() {
        produtosRecomendados.clear();
    }

    public void exibirRecomendacoes() {
        if (produtosRecomendados.isEmpty()) {
            System.out.println("Nenhuma recomendação disponível no momento.");
            return;
        }

        System.out.println("\n=== RECOMENDAÇÕES PERSONALIZADAS ===");
        Map<String, List<Produto>> recomendacoesPorCategoria = agruparPorCategoria();
        
        for (Map.Entry<String, List<Produto>> entry : recomendacoesPorCategoria.entrySet()) {
            System.out.println("\n--- " + entry.getKey() + " ---");
            for (Produto produto : entry.getValue()) {
                System.out.printf("Nome: %-30s | Descrição: %-40s | Preço: R$ %.2f%n",
                    produto.getNome(),
                    produto.getDescricao(),
                    produto.getPreco());
            }
        }
    }
} 